From e6944e019bb65aa062aea66f0d1848af7c6353c6 Mon Sep 17 00:00:00 2001
From: Tiejun Chen <tiejun.chen@windriver.com>
Date: Tue, 16 Aug 2011 09:40:13 +0800
Subject: [PATCH] pc-mc/fsl_p4080ds: disable QMan irq and schedule NAPI while reboot

When reboot guest OS without hardware reset we should disable QMan
IRQ and invoke NAPI to do a resource dump completely to avoid those
abnormal interrupt (and its associated invoked NAPI schedule) in
next boot phase.

Here we call dpaa_eth_napi_schedule() to implement this.

dpaa_eth_napi_schedule()
        |
        | /* Disable QMan IRQ and invoke NAPI */
        + qman_irqsource_remove(QM_PIRQ_DQRI)
        + napi_schedule(&percpu_priv->napi)

Signed-off-by: Tiejun Chen <tiejun.chen@windriver.com>
---
 drivers/net/dpa/dpaa_eth.c |   13 +++++++++++++
 1 files changed, 13 insertions(+), 0 deletions(-)

diff --git a/drivers/net/dpa/dpaa_eth.c b/drivers/net/dpa/dpaa_eth.c
index 4bbf9a0..3cd1f5b 100644
--- a/drivers/net/dpa/dpaa_eth.c
+++ b/drivers/net/dpa/dpaa_eth.c
@@ -2642,11 +2642,24 @@ static int __devexit __cold dpa_remove(struct of_device *of_dev)
 	struct device		*dev;
 	struct net_device	*net_dev;
 	struct dpa_priv_s	*priv;
+#if defined(CONFIG_PPC85xx_VT_MODE)
+	struct dpa_percpu_priv_s	*percpu_priv;
+#endif
 
 	dev = &of_dev->dev;
 	net_dev = dev_get_drvdata(dev);
 	priv = netdev_priv(net_dev);
 
+#if defined(CONFIG_PPC85xx_VT_MODE)
+	percpu_priv = per_cpu_ptr(priv->percpu_priv, raw_smp_processor_id());
+	/* When reboot guest OS without hardware reset we should disable QMan
+	 * IRQ and invoke NAPI to do a resource dump completely to avoid those
+	 * abnormal interrupt (and its associated invoked NAPI schedule) in
+	 * next boot phase.
+	 */
+	dpaa_eth_napi_schedule(percpu_priv);
+#endif
+
 	dev_set_drvdata(dev, NULL);
 	unregister_netdev(net_dev);
 
-- 
1.7.0.2

